src/sentry/utils/performance_issues/performance_detection.py

```diff
@@ -1395,6 +1403,202 @@ def _is_file_io_on_main_thread(self, span: Span) -> bool:
+    it transitions to the ContinuingMNPlusOne state.
+    """
+
+    __slots__ = ("settings", "recent_spans")
+
+    def __init__(
+        self, settings: Dict[str, Any], initial_spans: Optional[Sequence[Span]] = None
+    ) -> None:
+        self.settings = settings
+        self.recent_spans = initial_spans.copy() if initial_spans else []
+
+    def next(self, span: Span) -> Tuple[MNPlusOneState, Optional[PerformanceProblem]]:
+        # Can't be a potential MN+1 without at least 2 previous spans.
+        if len(self.recent_spans) <= 1:
+            self.recent_spans.append(span)
+            return (self, None)
+
+        # Has an MN pattern begun to repeat itself? If so, transition to the
+        # ContinuingMNPlusOne state.
+        for i, recent_span in enumerate(self.recent_spans[:-1]):
+            if self._equivalent(span, recent_span):
+                pattern = self.recent_spans[i:]
+                if self._is_valid_pattern(pattern):
+                    return (ContinuingMNPlusOne(self.settings, pattern, span), None)
+
+        # We haven't found a pattern yet, so remember this span and keep
+        # looking.
+        self.recent_spans.append(span)
+        max_sequence_length = self.settings["max_sequence_length"]
+        if len(self.recent_spans) > max_sequence_length:
+            self.recent_spans = self.recent_spans[1 : 1 + max_sequence_length]
+        return (self, None)
+
+    def _is_valid_pattern(self, pattern: Sequence[Span]) -> bool:
+        """A valid pattern contains at least one db operation and is not all equivalent."""
+        found_db_op = False
+        found_different_span = False
+
+        for span in pattern:
+            op = span.get("op") or ""
+            description = span.get("description") or ""
+            found_db_op = found_db_op or (
+                op.startswith("db") and description and not description.endswith("...")
+            )
+            found_different_span = found_different_span or not self._equivalent(pattern[0], span)
+            if found_db_op and found_different_span:
+                return True
+
+        return False
+
+
+class ContinuingMNPlusOne(MNPlusOneState):
+    """
+    The state for when we think we might have found a pattern: a sequence of
+    spans that has begun to repeat.
+
+    When the sequence is broken (either by a mismatched span or span iteration
+    finishing), returns to the SearchingMNPlusOne state, possibly returning a
+    PerformanceProblem if the detected sequence met our thresholds.
+    """
+
+    __slots__ = ("settings", "pattern", "spans", "times_repeated", "pattern_index")
+
+    def __init__(self, settings: Dict[str, Any], pattern: Sequence[Span], first_span: Span) -> None:
+        self.settings = settings
+        self.pattern = pattern
+
+        # The full list of spans involved in the MN pattern.
+        self.spans = pattern.copy()
+        self.spans.append(first_span)
+        self.times_repeated = 0
```

Comment:
Do you need to track `times_repeated`? It seems to be always a function of `len(self.spans)`

Reply:
Great call!